src/Microsoft.AspNet.Routing/Template/TemplateMatcher.cs

```diff
@@ -124,14 +120,14 @@ public class TemplateMatcher
                 }
                 else
                 {
-                    if (!MatchComplexSegment(routeSegment, requestSegment, Defaults, values))
+                    if (!MatchComplexSegment(routeSegment, requestSegment.ToString(), Defaults, values))
```

Comment:
I went ahead and did the substring here. We can optimize further if needed, but complex segments are fairly rare.